File: 2022/d4.py
#!/usr/bin/python3

Input = list[tuple[tuple[int, int], tuple[int, int]]]


def parse_input(lines: list[str]) -> Input:
    def rang_(atb):
        a, b = atb.split('-', 1)
        assert int(a) <= int(b)
        return int(a), int(b)

    def f1(a):
        l, _, r = a.partition(',')
        return rang_(l), rang_(r)
    return [f1(x) for x in lines if x]


def main1(inp: Input):
    total = 0
    for lr, rr in inp:
        a, b = lr
        c, d = rr
        if a <= c <= d <= b:
            total += 1
        elif c <= a <= b <= d:
            total += 1
    print('part1:', total)


def main2(inp: Input):
    total = 0
    for (a, b), (c, d) in inp:
        if b < c:
            ...
        elif d < a:
            ...
        else:
            total += 1
    print('part2:', total)


if __name__ == '__main__':
    lines = open('in4.txt').read().splitlines()
    inp = parse_input(lines)
    main1(inp)
    main2(inp)

File: 2022/src/d4.rs
use std::fs;

fn main() {
    let lines = fs::read_to_string("in4.txt").expect("failed to read input file");
    let input = lines.lines().map(parse_line).collect::<Vec<_>>();

    println!("part1: {}", solve1(&input));
    println!("part2: {}", solve2(&input));
}

fn parse_line(line: &str) -> [i32; 4] {
    let l;
    let r;
    if let Some((lp, rp)) = line.split_once(',') {
        l = lp;
        r = rp;
    } else {
        panic!("bad line");
    }

    fn parse_range(ab: &str) -> [i32; 2] {
        if !ab.contains('-') {
            return [-1, -1];
        }
        if let Some((lp, rp)) = ab.split_once('-') {
            let l = lp.parse::<i32>();
            let p = rp.parse::<i32>();
            if l.is_err() || p.is_err() {
                panic!("int parse");
            }
            [l.unwrap(), p.unwrap()]
        } else {
            [-1, -1]
        }
    }

    parse_range(l)
        .iter()
        .chain(&parse_range(r))
        .copied()
        .collect::<Vec<_>>()
        .try_into()
        .unwrap()
}

fn solve1(input: &[[i32; 4]]) -> u32 {
    let mut total = 0u32;
    for [a, b, c, d] in input {
        let mut f = (a..=b).contains(&c) && (a..=b).contains(&d);
        f = f || ((c..=d).contains(&a) && (c..=d).contains(&b));
        if f {
            total += 1;
        }
    }
    total
}

fn solve2(inp: &[[i32; 4]]) -> u32 {
    inp.iter()
        .map(|[a, b, c, d]| if b < c || d < a { 0 } else { 1 })
        .sum()
}
